### ISSUE TITLE
No error, bad output, from broken `fwd` signature

### ISSUE THREAD

**Issue by bprb** [shader-slang/slang-python#12]
Hello!

I solved my issue before hitting Submit, so instead may I suggest an improvement to the docs and/or error handling :^)

TLDR:
* I think it's useful to spell out a `fwd` example in the docs, do not underestimate my ability to mess it up ^_^
* Would it make sense to throw a compile error when using `[Differentiable]` with `TensorView` instead of `DiffTensorView`? 🤔 
   * This would also cause python code to throw an error when you don't pass a pair to `fwd`

I was getting bogus results trying to use `square.fwd` with the [documentation sample](https://shader-slang.com/slang/user-guide/a1-02-slangpy.html#getting-started-with-slangpy).

In particular, I called the `fwd` function with just a single input (thinking it just needs the original data, no dummy "ones") and just a single output (thinking I already get the answers from `square()`, so `square.fwd()` will give me the gradient):

```python
inputs = torch.tensor( (1, 2, 3, 4, 5, 6, 7, 8, 9, 10), dtype=torch.float).cuda()

squared = torch.zeros_like(inputs).cuda()
forward = torch.ones_like(inputs).cuda()

m.square(input=inputs, output=squared).launchRaw(blockSize=(32, 1, 1), gridSize=(64,1,1))
m.square.fwd(input=inputs, output=forward).launchRaw(blockSize=(32, 1, 1), gridSize=(64,1,1))

print(inputs)
print(squared)
print(forward)
```

Output:
```
tensor([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.], device='cuda:0')
tensor([  1.,   4.,   9.,  16.,  25.,  36.,  49.,  64.,  81., 100.], device='cuda:0')
tensor([  1.,   4.,   9.,  16.,  25.,  36.,  49.,  64.,  81., 100.], device='cuda:0')    # mild panic
```

This "unfortunately" worked because I added `[Differentiable]` but didn't change `TensorView` to `DiffTensorView` in the shader:
```hlsl
[AutoPyBindCUDA]
[CUDAKernel]
[Differentiable]
void square(TensorView<float> input, TensorView<float> output)
{
// ...
	output[dispatchIdx.x] = input[dispatchIdx.x] * input[dispatchIdx.x];
}
```

The result was that the generated code seemed to understand that I want a gradient, but actually redid the original calculation, which was a bit confusing 😛 

```cpp
__device__ void s_fwd_square_0(TensorView input_2, TensorView output_2)
{
    uint _S14 = (((blockIdx)) * ((blockDim)) + ((threadIdx))).x;
    uint _S15 = ((input_2).sizes[(0U)]);
    if(_S14 >= _S15)
    {
        return;
    }
    float _S16 = ((input_2).load<float>((_S14)));
    float _S17 = ((input_2).load<float>((_S14)));
    (output_2).store<float>((_S14), (_S16 * _S17));   // ?!
    return;
}

extern "C" {
__global__ void __kernel__square_fwd_diff(TensorView _S18, TensorView _S19)
{
    s_fwd_square_0(_S18, _S19);
    return;
}
```

Everything works with the correct shader...
```hlsl
[AutoPyBindCUDA]
[CUDAKernel]
[Differentiable]
void square(DiffTensorView<float> input, DiffTensorView<float> output)
{
	uint3 dispatchIdx = cudaBlockIdx() * cudaBlockDim() + cudaThreadIdx();
	if (dispatchIdx.x >= input.size(0))
		return;

	output[dispatchIdx.x] = input[dispatchIdx.x] * input[dispatchIdx.x];
}
```
... and pairing up both the input (all ones) and output (all zeros):
```python
inputs = torch.tensor( (1, 2, 3, 4, 5, 6, 7, 8, 9, 10), dtype=torch.float).cuda()
input_grad = torch.ones_like(inputs).cuda()

squared = torch.zeros_like(inputs).cuda()
forward = torch.ones_like(inputs).cuda()

m.square(input=inputs, output=squared).launchRaw(blockSize=(32, 1, 1), gridSize=(64,1,1))
m.square.fwd(input=(inputs, input_grad), output=(squared,forward)).launchRaw(blockSize=(32, 1, 1), gridSize=(64,1,1))
```

```
tensor([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.], device='cuda:0')
tensor([  1.,   4.,   9.,  16.,  25.,  36.,  49.,  64.,  81., 100.], device='cuda:0')
tensor([ 2.,  4.,  6.,  8., 10., 12., 14., 16., 18., 20.], device='cuda:0')   # yeay!
```

So, in the end, simple user error -- but if there is no use case where mixing Differentiable with plain TensorViews makes sense, then perhaps it can be detected and disallowed?

Thanks!
bert

**Comment by saipraveenb25** [comment]
Thank you for bringing it to our attention!

This is unfortunately a tricky situation. Since we do want to allow regular `TensorView` types for passing non-differentiable data, it would be confusing to throw errors/warnings when using `TensorView` with a `[Differentiable]` function. 

As you suggested, for now, we'll add an example with `.fwd()` to the documentation to make the interface clearer, i.e., `.fwd()` computes both the regular output & its derivative

**Comment by bprb** [comment]
Hi Sai, thank you for the quick reply!  I see, this is a valid thing to do.  We'll just have to be careful then :)
Thanks for taking the suggestion about the docs!
